matrix/src/armory/matrix/matrix.py

```diff
@@ -0,0 +1,267 @@
+"""Matrix generation utilities"""
+
+import concurrent.futures
+from copy import deepcopy
+from typing import Any, Callable, Dict, Iterable, List, Mapping, Optional, Tuple
+
+
+def product(
+    prior: Dict[str, Any], remaining: List[Tuple[str, Any]]
+) -> Iterable[Mapping[str, Any]]:
+    """
+    Generates a cartesian product of parameters.
+
+    This function works recursively, iterating over each parameter range in a
+    depth-first manner. We do this rather than use `itertools.product` in order
+    to support dynamic, or dependent, parameter ranges that are created based
+    on the preceding parameter values.
+
+    Args:
+        prior: Key-value parameter mapping of parameters generated so far
+        remaining: List of remaining key and value-iterable pairs
+
+    Yields:
+        Key-value parameter mappings for each row in the cartesian product
+        matrix
+    """
+    if not remaining:
+        yield prior
+    else:
+        key, values = remaining[0]
+        if callable(values):
+            values = values(**prior)
+        if not isinstance(values, Iterable):
+            values = [values]
+        for value in values:
+            next = deepcopy(prior)
```

Comment:
It would be useful to think through carefully if this really needs to be a deep copy or would a shallow copy suffice?
What guarantees are being made if a "matrixed" function is re-evaluated?

Reply:
In a perfect case, the matrixed function doesn't modify any of its inputs, so a shallow copy is all that's needed. I also expect that for 90% or more of use-cases, the parameters are going to be simple values (numbers, bools, strings, function pointers perhaps), and I'm not sure what the performance difference is for cases where deepcopy and shallow copies have the same result.

A shallow copy is _probably_ fine.